File: 01_Easy/0014_Longest_Common_Prefix.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        string common_prefix = "";
        string test_prefix = "";
        string curr_prefix = "";

        string comparing_string = strs.at(0);
        //char next_char = comparing_string.at(0);
        bool flag = false;
        for (int i = 1; i <= comparing_string.length(); i++){
            test_prefix = comparing_string.substr(0, i);

            for (int j = 1; j < strs.size(); j++){
                curr_prefix = strs.at(j).substr(0, i);

                flag = (curr_prefix != test_prefix);
                if (flag) {break;}
            }
            if (flag) {break;}
            common_prefix = test_prefix;
        }
        return common_prefix;
    }
        //std::cout << test_prefix << '\n';
};

File: 01_Easy/0014_Longest_Common_Prefix.py
class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:

        # Sort the array by word size
        strs.sort(key=len)

        common_prefix = ""
        longest_prefix = ""

        # Edge cases
        if len(strs) == 1:
            return strs[0]
        if len(strs) == 0:
            return ""

        # Main loops
        for j in range(len(strs[0])):
            common_prefix = strs[0][0:j+1]
            for i in range(1, len(strs)):
                prefix = strs[i][0:j+1]
                if prefix != common_prefix:
                    return longest_prefix
            longest_prefix = strs[0][0:j+1]

        return longest_prefix